### PR TITLE
Use name and type string from struct for keycloak oidc & oidc

### PR DESCRIPTION
**Pull request by brendarearden**
**Issue(s)**
https://github.com/rancher/rancher/issues/32990
https://github.com/rancher/rancher/issues/32941
https://github.com/rancher/rancher/issues/32936

**Problem**
The type and name are not being set properly with `.GetName()` and `.GetType()` which is resulting in both Keycloak OIDC and OIDC providers to become enabled. Both of these being enabled at the same time causes

1. The login screen not to know which auth provider to show to the user
2. Once they become enabled, there is no way to disable them because the disable doesn't have the correct Name of the provider to disable it. 

**Solution**
Set the name and type values on the OIDCProvider struct so that the values are correct at time of configuration and it doesn't rely on inheritance to call the correct function.

### CODE REVIEW

File: pkg/auth/providers/oidc/oidc_provider.go [hunk 1]
```diff
@@ -255,10 +255,10 @@ func (o *OpenIDCProvider) userToPrincipal(userInfo *oidc.UserInfo, info claimInf
 		displayName = userInfo.Email
```
Comment (cbron) [inline pkg/auth/providers/oidc/oidc_provider.go:255]: Nit: use arg variable name of `claiminfo` instead of `info`. As both those variables have info in their name.

File: pkg/auth/providers/oidc/oidc_actions.go [hunk 2]
```diff
@@ -42,7 +42,7 @@ func (o *OpenIDCProvider) ActionHandler(actionName string, action *types.Action,
 func (o *OpenIDCProvider) ConfigureTest(actionName string, action *types.Action, request *types.APIContext) error {
 	//verify body has all required fields
 	input, err := handler.ParseAndValidateActionBody(request, request.Schemas.Schema(&managementschema.Version,
-		o.getType()))
+		o.Type))
```
Comment (cbron) [inline pkg/auth/providers/oidc/oidc_actions.go:45]: So this all gets inherited by keycloak OIDC ? And therefore O is of keycloak type.

File: pkg/auth/providers/oidc/oidc_provider.go [hunk 3]
```diff
@@ -169,11 +169,11 @@ func (o *OpenIDCProvider) SearchPrincipals(searchValue, principalType string, to
...
 
 	p := v3.Principal{
-		ObjectMeta:    metav1.ObjectMeta{Name: o.GetName() + "_" + principalType + "://" + searchValue},
+		ObjectMeta:    metav1.ObjectMeta{Name: o.Name + "_" + principalType + "://" + searchValue},
 		DisplayName:   searchValue,
 		LoginName:     searchValue,
 		PrincipalType: principalType,
-		Provider:      o.GetName(),
+		Provider:      o.Name,
 	}
```
Comment (cbron) [inline pkg/auth/providers/oidc/oidc_provider.go:177]: Would be nice to re-use the `toPrinciples` methods below but might be too complicated here since it looks like you are using both and you don't have necessary info. 
Unrelated to this PR, but I am confused why you just return this directly with the search value included, its just a placeholder ?

**Review by paynejacob — DISMISSED** [review]

File: pkg/auth/providers/oidc/oidc_actions.go (hunk 2 above)
Comment (brendarearden) [reply pkg/auth/providers/oidc/oidc_actions.go:45]: Yes, the `o` at this point is the `OpenIDCProvider` on the `keyCloakOIDCProvider` struct. This issue that this fixes was  was when functions in the OpenIDCProvider interface were being called, it had no idea about the "overriden"  `o.getType()` that I had created in `keyCloakOIDCProvider` so it wasn't returning OpenIDCProvider values and not the keyCloakOIDCProvder values. This meant some types and names were getting set wrong ultimately leading to both providers getting enabled with the rest of the problems that come with it. By setting the value as string, it ensures the value will be available regardless of if its called in the interface or if I change the func definition in the `keyCloakOIDCPRovider` struct. There could be better ways to resolve this, but this seemed like a good solution for now.

File: pkg/auth/providers/keycloakoidc/keycloak_provider.go [hunk 4]
```diff
@@ -87,10 +85,10 @@ func (k *keyCloakOIDCProvider) toPrincipal(principalType string, acct account, t
 		displayName = acct.Username
 	}
 	princ := v3.Principal{
-		ObjectMeta:  metav1.ObjectMeta{Name: k.GetName() + "_" + principalType + "://" + strconv.Itoa(acct.ID)},
+		ObjectMeta:  metav1.ObjectMeta{Name: k.Name + "_" + principalType + "://" + strconv.Itoa(acct.ID)},
 		DisplayName: displayName,
 		LoginName:   acct.Username,
-		Provider:    k.GetName(),
+		Provider:    k.Name,
```
Comment (prachidamle) [inline pkg/auth/providers/keycloakoidc/keycloak_provider.go:91]: k.GetName() returns the Name itself, so why this change?

File: pkg/auth/providers/oidc/oidc_provider.go (hunk 3 above)
Comment (prachidamle) [reply pkg/auth/providers/oidc/oidc_provider.go:177]: I think this is because  OIDC does not support searching?

File: pkg/auth/providers/keycloakoidc/keycloak_provider.go (hunk 4 above)
Comment (brendarearden) [reply pkg/auth/providers/keycloakoidc/keycloak_provider.go:91]: I changed it here to match how I am using Name and Type in the oidc_provider. But I do like the idea of using GetName() in the specific providers so I will revert this.

File: pkg/auth/providers/oidc/oidc_provider.go (hunk 1 above)
Comment (brendarearden) [reply pkg/auth/providers/oidc/oidc_provider.go:255]: updated `info` to `claimInfo` to match the `userInfo` variable

File: pkg/auth/providers/oidc/oidc_provider.go (hunk 3 above)
Comment (brendarearden) [reply pkg/auth/providers/oidc/oidc_provider.go:177]: Correct - This is how our SAML providers are currently handling `SearchPrincipals` because they do not support search functionality for the various SAML providers.

The other option would be a `return nil, errors.New("[generic oidc]: not implemented")` instead of implementing it.

File: pkg/auth/providers/oidc/oidc_actions.go (hunk 2 above)
Comment (cbron) [reply pkg/auth/providers/oidc/oidc_actions.go:45]: I had to brush up my memory here to remember: https://play.golang.org/p/IGMYjG9mQbQ

> Embedding types introduces the problem of name conflicts but the rules to resolve them are simple. First, a field or method X hides any other item X in a more deeply nested part of the type. If log.Logger contained a field or method called Command, the Command field of Job would dominate it.

https://golang.org/doc/effective_go#embedding

If you added a keycloak.printer there it would only return that one. So in reverse if you dropped the GetType from Keycloak it would refer back to the OIDC one. 

Using strings avoids this whole confusion.

**Review by cbron — APPROVED** [review]
lgtm

**Review by prachidamle — APPROVED** [review]
LGTM
